signer.go

```diff
@@ -36,8 +40,86 @@ func hashSortedBigInt(lst []string) *big.Int {
 }
 
 var goproxySignerVersion = ":goroxy1"
+var certprivRSA crypto.Signer
+var certprivECDSA crypto.Signer
+var crtCache = cache.New(20*time.Minute, 10*time.Minute)
+var serialCA uint64 = 9999
 
 func signHost(ca tls.Certificate, hosts []string) (cert *tls.Certificate, err error) {
+	key := strings.Join(hosts[:], "/")
+	value, exist := crtCache.Get(key)
+	if exist {
+		cert = value.(*tls.Certificate)
+		return
+	}
+
+    cert, err = generateCertificate(ca, hosts)
+
+	crtCache.Add(key, cert, cache.DefaultExpiration)
```

Comment:
Race condition scenario:

Two concurrent calls to signHost for hostname with not yet cached crt.
The second call will fail to add cache, as it is existing, although it didn't exist when line #50 was called.

Reply:
@engineering-this: how would you solve that race condition?

Reply:
I'd go with a semaphore.

```go
var semaphore = make(chan struct{}, 1)

func signHost(ca tls.Certificate, hosts []string) (cert *tls.Certificate, err error) {
	semaphore <- struct{}{}
	defer func() {
		<-semaphore
	}()
	//...
```